plugins/amqp_trx_plugin/amqp_trx_plugin.cpp

```diff
@@ -0,0 +1,175 @@
+
+} // anonymous
+
+namespace eosio {
+
+using boost::signals2::scoped_connection;
+
+struct amqp_trx_plugin_impl : std::enable_shared_from_this<amqp_trx_plugin_impl> {
+
+   chain_plugin* chain_plug = nullptr;
+   amqp_trace_plugin* trace_plug = nullptr;
+   // use thread pool even though only one thread currently since it provides simple interface for ioc
+   std::optional<eosio::chain::named_thread_pool> thread_pool;
+   std::optional<amqp> amqp_trx;
+
+   std::string amqp_trx_address;
+   std::atomic<uint32_t> trx_in_progress_size{0};
+
+   // called from amqp thread
+   bool consume_message( const char* buf, size_t s ) {
+      try {
+         fc::datastream<const char*> ds( buf, s );
+         fc::unsigned_int which;
+         fc::raw::unpack(ds, which);
+         if( which == fc::unsigned_int(transaction_msg::tag<chain::packed_transaction_v0>::value) ) {
+            chain::packed_transaction_v0 v0;
+            fc::raw::unpack(ds, v0);
+            auto ptr = std::make_shared<chain::packed_transaction>( std::move( v0 ), true );
+            handle_message( std::move( ptr ) );
+         } else if ( which == fc::unsigned_int(transaction_msg::tag<chain::packed_transaction>::value) ) {
+            auto ptr = std::make_shared<chain::packed_transaction>();
+            fc::raw::unpack(ds, *ptr);
+            handle_message( std::move( ptr ) );
+         } else {
+            FC_THROW_EXCEPTION( fc::out_of_range_exception, "Invalid which ${w} for consume of transaction_type message", ("w", which) );
+         }
+         return true;
+      } FC_LOG_AND_DROP()
+      return false;
+   }
+
+private:
+
+   // called from amqp thread
+   void handle_message( chain::packed_transaction_ptr trx ) {
+      const auto& tid = trx->id();
+      dlog( "received packed_transaction ${id}", ("id", tid) );
+
+      auto trx_in_progress = trx_in_progress_size.load();
+      if( trx_in_progress > def_max_trx_in_progress_size ) {
+         wlog( "Dropping trx, too many trx in progress ${s} bytes", ("s", trx_in_progress) );
+         if( trace_plug ) {
+            std::string err = "Dropped trx, too many trx in progress " + std::to_string( trx_in_progress ) + " bytes";
+            trace_plug->publish_error( trx->id().str(), chain::tx_resource_exhaustion::code_enum::code_value, std::move(err) );
+         }
+         return;
+      }
+
+      trx_in_progress_size += trx->get_estimated_size();
+      app().post( priority::medium_low, [my=shared_from_this(), trx{std::move(trx)}]() {
+         my->chain_plug->accept_transaction( trx,
+            [my, trx](const fc::static_variant<fc::exception_ptr, chain::transaction_trace_ptr>& result) mutable {
+               // publish to trace plugin as execptions are not reported via controller signal applied_transaction
+               if( result.contains<chain::exception_ptr>() ) {
+                  auto& eptr = result.get<chain::exception_ptr>();
+                  if( my->trace_plug ) {
```

Comment:
I get a segfault if the `amqp_trace_plugin` was not enabled.  I think the reason is that `my->trace_plug` is never null even if `amqp_trace_plugin` skipped initialization.

Reply:
Interesting the segfault seems to be a macos only thing. Anyway, now check if initialized.

Reply:
I don't like the fact that the behavior changes depending on the order that the plugins are specified on the command line.  Would there be any problem with checking `my->trace_plug->get_state() == started` here instead of in `plugin_startup`?

Reply:
I think this still leaves a window on shutdown where the trx plugin is running, but the trace plugin is not.  This looks mostly harmless to me.
- The trace will be dropped silently (which can already happen at startup and shutdown, anyway)
- The trx_plugin_impl object will never be destroyed, because the callback holds a `shared_ptr` to it.  (We're already shutting down, so memory leaks are not critical)

Reply:
Seems like that kind of issue should be handled in `chain_plugin`  and/or `application::exec()` as all plugins which hook up to signals have that issue.

Reply:
My point was all plugins detach from the signals but the application thread can still be processing so they can miss the signal call.

Reply:
I was proposing that since every plugin depends on `chain_plugin` and `chain_plugin` is soon to be a library instead of a "plugin" that `chain_plugin` should hold the invariant that by the time it shuts down the application thread has processed all queued tasks and that no controller signals will be emitted. Then all plugins can safely disconnect from signals knowing they will not miss anything. 
There would still be a race on posting to `app().post()` for plugins like `amqp_trx_plugin` and `net_plugin` and `http_plugin`. However, not having your trx processed seems like less of a problem then it being processed and put into a block but missing the signals.

Reply:
Right.  I'm not really concerned about the lost notifications right now, as it's a more general problem and we're not making it worse.  I'm somewhat more concerned about the `shared_ptr` created in `amqp_trx_plugin_impl::publish_error` preventing the `amqp_trx_plugin_impl` from being destroyed.

Reply:
Not sure how that could be since: "Uninvoked handler objects that were scheduled for deferred invocation on the io_context, or any associated strand, are destroyed."

Reply:
I take it you're looking at https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/reference/io_context/_io_context.html.  The thing is, the relevant io_context is not destroyed either, because it's a member of the `amqp_trace_plugin_impl` (I meant `trace_plugin` not `trx_plugin` above ).

Reply:
Ah, ok, I was looking at the wrong thing. I see what  you mean now.